presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestMaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,394 @@
+        Query expectedViewQuery = (Query) sqlParser.createStatement(expectedViewSql);
+
+        assertEquals(optimizedBaseToViewQuery, expectedViewQuery);
+    }
+
+    @Test
+    public void testQueryOptimizationUsingMaterializedViewWithOrderBy()
+    {
+        SqlParser sqlParser = new SqlParser();
+        String baseTable = "base_table";
+        String view = "view";
+
+        // Definition of materialized view
+        String originalViewSql = format("SELECT a, b, c From %s", baseTable);
+        Query originalViewQuery = (Query) sqlParser.createStatement(originalViewSql);
+
+        String baseQuerySql = format("SELECT a, b, c From %s ORDER BY c ASC, b DESC, a", baseTable);
+        Query baseQuery = (Query) sqlParser.createStatement(baseQuerySql);
+
+        Table viewTable = new Table(QualifiedName.of(view));
+
+        Query optimizedBaseToViewQuery = (Query) new MaterializedViewQueryOptimizer()
+                .process(baseQuery, new MaterializedViewQueryOptimizer.MaterializedViewQueryOptimizerContext(viewTable, originalViewQuery));
+
+        String expectedViewSql = format("SELECT a, b, c From %s ORDER BY c ASC, b DESC, a", view);
+        Query expectedViewQuery = (Query) sqlParser.createStatement(expectedViewSql);
+
+        assertEquals(optimizedBaseToViewQuery, expectedViewQuery);
+    }
+
+    @Test
+    public void testQueryOptimizationUsingMaterializedViewWithOrderByWithAlias()
+    {
+        SqlParser sqlParser = new SqlParser();
+        String baseTable = "base_table";
+        String view = "view";
+
+        // Definition of materialized view
+        String originalViewSql = format("SELECT a as mv_a, b, c as mv_c From %s", baseTable);
+        Query originalViewQuery = (Query) sqlParser.createStatement(originalViewSql);
+
+        String baseQuerySql = format("SELECT a, b, c From %s ORDER BY c ASC, b DESC, a", baseTable);
+        Query baseQuery = (Query) sqlParser.createStatement(baseQuerySql);
+
+        Table viewTable = new Table(QualifiedName.of(view));
+
+        Query optimizedBaseToViewQuery = (Query) new MaterializedViewQueryOptimizer()
+                .process(baseQuery, new MaterializedViewQueryOptimizer.MaterializedViewQueryOptimizerContext(viewTable, originalViewQuery));
+
+        String expectedViewSql = format("SELECT mv_a, b, mv_c From %s ORDER BY mv_c ASC, b DESC, mv_a", view);
+        Query expectedViewQuery = (Query) sqlParser.createStatement(expectedViewSql);
+
+        assertEquals(optimizedBaseToViewQuery, expectedViewQuery);
+    }
+
+    @Test
+    public void testQueryOptimizationUsingMaterializedViewWithFunctionCallInOrderBy()
+    {
+        SqlParser sqlParser = new SqlParser();
+        String baseTable = "base_table";
+        String view = "view";
+
+        // Definition of materialized view
+        String originalViewSql = format("SELECT COUNT(a) as mv_count_a, b From %s GROUP BY b", baseTable);
```

Comment:
counts are complicated, lets skip count examples. I would prefer to have few WITH clause examples.